### ISSUE TITLE
Pyenv shim spam breaks $PATH, scripts

### ISSUE THREAD

**Issue by deanishe** [pyenv/pyenv#1112]
In addition to shadowing—and therefore breaking—programs installed on the system `$PATH` with its shims, pyenv also breaks `if hash progname; then ...` tests.

If a program isn't installed system-wide, but *is* installed in a venv, the pyenv shim causes the test to erroneously pass, and of course subsequently running the program results in a "hurr durr, this command exists in these Python versions…" error message when a pyenv shim gets run instead of the real program.

The former issue (not calling existing programs) can be worked around with an extension, but AFAIK the latter can't.

Is there a solution to this problem?

pyenv is darn handy for managing venvs and Python versions. A completely borked `$PATH` is *not* a price worth paying, however.

**Comment by yyuu** [comment]
Generally speaking, the usage of shims is the core design of pyenv and it'll prevent the user to the availability of some command by checking presence of the command executable. This is core design issue and there's no easy and general workaround for it. One idea would be invoking some no-op command like `--version` (e.g. `pip --version`) to verify the availability of the command.

**Comment by deanishe** [comment]
Thanks for the response.

> This is core design issue and there's no easy and general workaround for it

Then what needs to change to stop it being a core design issue? Sure, it's possible to work around in most situations, but programs shouldn't be breaking `$PATH` in the first place.

(Any pyenv *does* break `$PATH`: it deliberately breaks the behaviour of they system and other programs by shadowing existing and non-existing real programs with shim scripts that throw errors.)

Maybe I've missed something due to the way I use pyenv (as a better virtualenv), but I don't really understand why pyenv even needs to drop its shims in `$PATH` in the normal case. Their default function ("I look like a real program, but I'm not! You can find one here or here") seems like a command-line version of ISPs hijacking failed DNS lookups and showing their a webpage instead of returning `NXDOMAIN`: it is helpful in certain very limited circumstances, but mostly it just breaks shit because the system no longer works the way literally everything else expects it to.

pyenv can fiddle with `$PATH` when I activate/deactivate a Python version/virtualenv. It should (have an option to) leave `$PATH` the hell alone when I'm not actively using pyenv.

**Comment by bitfinity** [comment]
Yes, I'm trying to use the system python. But python is mapping to ~/.pyenv/shims/python even though I've run `pyenv deactivate`.

**Comment by blueyed** [comment]
The issue is that the shims have to be in the $PATH to be found, so that it can delegate to the selected version (in the case where it is installed/available).
(pyenv by now should at least exit with code 127 when a version is not installed.)

> pyenv can fiddle with $PATH when I activate/deactivate a Python version/virtualenv. It should (have an option to) leave $PATH the hell alone when I'm not actively using pyenv.

A solution to this might be using shell functions instead of shim scripts, i.e. generating a shell function for each entry in "shims", and not adding "shims" to $PATH.

However, this would make pyenv only work in interactive shells, and not e.g. for programs (like Neovim etc) that look up / use e.g. "python" as a binary.

**Comment by deanishe** [comment]
> The issue is that the shims have to be in the $PATH to be found, so that it can delegate to the selected version (in the case where it is installed/available).

Why? I'm not a heavy user of pyenv, so presumably I'm missing something.

It seems to me that the only *useful* purpose of a shim is to call a program in the ~/bin directory of the active Python version. Why not just add that directory to $PATH when I activate the corresponding Python version instead of creating a bunch of shims?

If it must be shims, why can't they be adjusted when Python versions are (de)activated so that only working shims exist at any given time?

Making it appear that a program is available when it isn't and overriding working programs with do-nothing shims is pretty bad behaviour.

**Comment by blueyed** [comment]
> Why not just add that directory to $PATH when I activate the corresponding Python version instead of creating a bunch of shims?

Once you use pyenv, all versions are activated according to this.
You might be better off by only enabling pyenv on demand then, too - e.g. by creating a function yourself that does the init when you start using it (I can give you pointers).
It needs special handling for `pyenv local` then though for example, but there are other ways to activate virtualenvs, e.g. https://github.com/Tarrasch/zsh-autoenv/#automatically-activate-python-virtualenvs.

> If it must be shims, why can't they be adjusted when Python versions are (de)activated so that only working shims exist at any given time?

That is the case already - there should only be shims for programs (including `python`) that are installed, and those get rehashed / cleaned up (via `pyenv rehash`).

> Making it appear that a program is available when it isn't and overriding working programs with do-nothing shims is pretty bad behaviour.

That is only the case for when foo is installed in some virtualenv (and a shim for it has been created therefore).

Thinking about it however, I could imagine it to be smarter indeed, e.g. by using some extra layer, where symlinks would be created to shims that actually exist (reminds me of NixOS/nixpkgs).
So `pyenv global`, `pyenv local` and `pyenv shell` would then trigger creating / managing this.
It might also be possible already without the extra layer, but only having shims for programs that exist.
I've noticed however that pyenv-virtualenv explicitly creates shims for all virtualenvs (similar to pyenv itself creating it for all versions).

Some var or hook could be added (e.g. when a new version is detected, which is only the case when calling the actual shim however with `pyenv local` (and likely also `pyenv global`), since only then the config is read.
So it might only be feasible with `pyenv shell` currently, but at least for `pyenv local` it could use the shell's `cd` hook to detect if there is a new version then.

**Comment by blueyed** [comment]
/cc @mislav 
(But likely it is better to invest time into https://github.com/mislav/anyenv instead.)

**Comment by blueyed** [comment]
One caveat at least: every shell would need a separate shims directory (e.g. based on its PID, but better some UUID/random string), which gets set with `init`, or the shims entry/placeholder in `$PATH` could be replaced when the version is changed.
Given that the shims for different versions are rather static, it would make sense to have a separate dir  for them (e.g. based on their name), and then link to this for the different shell's $PATH.

**Comment by deanishe** [comment]
> there should only be shims for programs (including python) that are installed

I didn't say "installed", I said "working". At any given time, a bunch of the shims will throw errors because they belong to non-activated versions. This breaks $PATH and other software.

> That is only the case for when foo is installed in some virtualenv

If that version isn't activated, the shim throws an error. As such, it doesn't much matter that foo is indeed installed somewhere: pyenv's foo shim is breaking $PATH as long as the *active version* doesn't contain foo.

> but there are other ways to activate virtualenvs

Activation is easy. I really like pyenv because of its ability to cleanly install different versions of Python and create virtualenvs from them.

> Thinking about it however, I could imagine it to be smarter indeed, e.g. by using some extra layer

I know nothing about pyenv's internals. I would just like my $PATH—at any given time—to be free of shims that literally throw "command not found" errors.

These, simply put, should not exist. They break $PATH by causing things like `which python3.5` to incorrectly succeed or by overriding actual working versions later on $PATH.

**Comment by mislav** [comment]
I'm not a user of pyenv, but I'm the maintainer of rbenv, which pyenv is a direct derivative of. I agree that the core design of rbenv has this big flaw in some use cases. However, in most other use cases, this “flaw” is actually an important design feature.

1. Let's suppose that you have a system installation of Ruby `/usr/local/bin/ruby` and a system installation of Bundler (important tool for managing Ruby dependencies `/usr/local/bin/bundle`).

2. Now, you add rbenv to this system, install a Ruby 2.3.4, and install its own Bundler. You would end up with `~/.rbenv/shims/bundle` that activates `~/.rbenv/versions/2.3.4/bin/bundle` when `bundle` is invoked in a context of Ruby 2.3.4.

3. Now, add another version of Ruby: 2.4.0. However, let's say you _forgot to install_ Bundler on this version. Running `bundle` in this version would error with: “The 'bundle' command exists in these Ruby versions: 2.3.4”. Notice how this call **doesn't** get dispatched to `/usr/local/bin/bundle`, which would activate Bundler for a wrong (system) Ruby version and mess up this project's dependencies. **The shadowing of executables through `~/.rbenv/shims` is an important design feature** meant to scope all calls to Ruby tools to a single Ruby version, and fail otherwise.

This feature isn't without its flaws. If a single Ruby gem installed to any of the rbenv-managed Ruby versions were to add a `man` executable, then `~/.rbenv/shims/man` would be generated and that would shadow the system `man` on any invocation, even when the Ruby version that contains the offending gem isn't activated. This scenario happened to people using `yarn` (a tool for managing JavaScript dependencies).

I like [chruby's](https://github.com/postmodern/chruby) design philosophy, where the currently activated Ruby simply adds its `bin` directory to PATH. This is dependent on an interactive shell and cd hacks, though. Rbenv's philosophy depends on the PATH lookup mechanism and not the presence of an interactive shell. They are different approaches. They have their drawbacks and benefits.

@deanishe I'm not sure what you're expecting with this issue. Are you suggesting a whole redesign of the core philosophy of how this version manager works. If so, would you suggest an alternative?

> Then what needs to change to stop it being a core design issue?

A fundamental rewrite of the whole implementation, and a rollout to all existing users that would somehow need to preserve backwards compatibility, but would probably fail doing so (because of the aforementioned rewrite).

Or am I missing a simpler solution?

**Comment by deanishe** [comment]
@mislav Thanks for the explanation.

> However, in most other use cases

In *one* use case. When you're currently writing Python code. That's the only time command-not-found shims are in any way useful. The rest of the time, they're just breaking $PATH.

Personally, I find using `autoenv` or a `.python-version` to activate pyenv's versions more than good enough to ensure the right version is activated and the right `pip` (or whatever) gets run.

Ideally, I would like a global "I accept responsibility for running the right program; don't break my $PATH" setting, which would tell pyenv not to create any command-not-found shims that will fail instead of running the corresponding program (i.e. only shims for programs in the active version exist).

At the least, I would like pyenv to not create failing shims when it's "off", i.e. I have it set to "system" version. In this situation, the whole "shadowing prevents you from running the wrong version" argument seems moot.

I spend a small fraction of my time coding Python, and I would like to not have a broken $PATH the rest of the time.

> A fundamental rewrite of the whole implementation, and a rollout to all existing users that would somehow need to preserve backwards compatibility

Again, I'm not familiar with pyenv's internals, but I'd be very surprised if not creating shims for the non-active version would require a "fundamental rewrite of the whole implementation".

As for backwards-compatibility, keep the current behaviour as default and add an option.

**Comment by blueyed** [comment]
I have experimented a bit with separate shim dirs per version: https://github.com/pyenv/pyenv/pull/1185

**Comment by techalchemy** [comment]
Just to add -- I've installed an anaconda distribution which is not activated, and I can therefore no longer run `tar -xjvf`, `pandoc` and countless other commands which come with anaconda and are therefore symlinked into pyenv's shim directory.  I am a maintainer of pipenv which provides installation functionality for pyenv directly, but seeing that it breaks my basic workflows I'm not really sure how to move forward. I can't even delete the shims as they are recreated every time I re-source my config file.  This is a pretty significant interruption for something that has been dismissed as an intentional design decision but which actually just fundamentally breaks basic system commands like `bzip2` without offering any passthrough or entertaining options for a solution.

For now I am just removing the shims entirely until I can determine what to do with this moving forward

**Comment by blueyed** [comment]
@techalchemy 
Please consider giving #1185 a try.

**Comment by anentropic** [comment]
I think I have this issue

I have installed [powerline-shell](https://github.com/b-ryan/powerline-shell) in my system python

Every time I activate a pyenv virtualenv (or pip install/uninstall a package in the venv after activating) it recreates a useless shim file which causes the error `pyenv: powerline-shell: command not found` and breaks my shell prompt

I then have to locate the shim and delete it for everything to work again
(alternatively, installing it into my venv also works, but I really don't want to introduce it as a project dependency everywhere)

Should I have installed powerline-shell differently?  I would like it to work when I have no venv active as well as when I do.

This is so annoying, any advice or workaround much appreciated

**Comment by leoalekseyev** [comment]
I switched from using conda to manage Python versions to pyenv. However, for some scientific computing projects it's easiest to use Anaconda; I installed Anaconda under pyenv and ran into this issue. It's extremely frustrating, and I'm planning to go back to conda until it's fixed.  Perhaps there should be a warning in FAQ or somewhere that pyenv doesn't play well with Anaconda.

**Comment by deanishe** [comment]
@techalchemy [pyenv-which-ext](https://github.com/pyenv/pyenv-which-ext) should prevent the shims from throwing errors for programs that *are* installed elsewhere, such as `git`.

The main problem is that the existence of the shims causes tests like `if hash progname; then ...` to succeed when actually running the command will get you a "command not found" error from the shim.

Any solution to that problem would presumably require the removal of the error-throwing shims from $PATH, however, which would also solve the shadowing problem and make the above extension redundant.

**Comment by 1oglop1** [comment]
I wish this was finally fixed and shims were separated by versions folders or else.
I have another issue caused by this problem:
It is really annoying especially when I'm working in pycharm.
If PyCharm is started from one terminal with `charm .` and then I change directory to different terminal where the python version is different but same command is installed and I run `charm .` the shims are preventing me from executing the command from the second virtual environment due to fact that it's already available..

eg:

venv1_python36: `cmd --version` -> `v1.0` (`charm .` -> pycharm terminal same effect)
venv2_python3.7 `cmd --version -> `v2.0` (`charm .` (will launch second instance) -> pycharm terminal `v1.0`)

**Comment by blueyed** [comment]
You are free to try and provide feedback with https://github.com/pyenv/pyenv/pull/1185.

**Comment by 1oglop1** [comment]
@blueyed I tried that version but it was bit too much work to get it working, cuz I wasn't aware how much is abstracted via `brew`. But after I got it working it seemed to behave just fine

**Comment by matthewatabet** [comment]
It would be great if pyenv only added the shims location when a venv is activated. I ran into this shadowing problem and it's very difficult to resolve from a sysadmin perspective. Thanks!

**Comment by 1oglop1** [comment]
@matthewatabet could you try #1185 and provide some feedback as well?

**Comment by 1oglop1** [comment]
@blueyed What is the easiest way to keep your fork up to date with main repo + some easier installation on Mac?

**Comment by ryudice** [comment]
This makes pyenv completely useless, it causes a lot of erros when you install anaconda, because anaconda comes with its own gcc and make executables. Tons of time wasted due to this.

**Comment by mazzma12** [comment]
> This makes pyenv completely useless, it causes a lot of erros when you install anaconda, because anaconda comes with its own gcc and make executables. Tons of time wasted due to this.

Adding some context to that description from my experience, the failure indeed comes with mixing anaconda and regular environment managed with pip. Gor example, let's take gdal package which a geo library available via `apt` and also with `conda`, but not with `pip`.

For example `conda install rasterio` will install its own version of GDAL in `env1` managed by conda.
But when calling gdal from `env2` it will detect that gdal is available in `env1` and will fail.
In my understanding, it shall fallback to the system-wide gdal

**Comment by deanishe** [comment]
@blueyed How does #1185 work?

I've switched my pyenv install to your branch, but ~/.pyenv/shims is still full of `$PATH`-breaking "command not found" shims? Is there something I have to do to make them go away?

**Comment by tmzhuang** [comment]
Not too sure if this is related or if this is even something to do with pyenv. I'm using powerline and there's some funky stuff going on with the the way the path is displayed depending on if I have a local pyenv set.

![image](https://user-images.githubusercontent.com/1159725/63188842-84c55980-c030-11e9-82b7-3d185b647bf3.png)
Note the virtualenv is incorrectly displayed until a new command is made. Also cd'ing to a different directory does not update the path correctly. This is not the case if I'm not using a .python-version to start a virtualenv.
![image](https://user-images.githubusercontent.com/1159725/63189003-e84f8700-c030-11e9-8943-d38cf76024fd.png)
Although it does look like there are issues updating the virtualenv as well.

**Comment by blueyed** [comment]
@deanishe 
> @blueyed How does #1185 work?
> 
> I've switched my pyenv install to your branch, but ~/.pyenv/shims is still full of `$PATH`-breaking "command not found" shims? Is there something I have to do to make them go away?

I think you need to remove the shims on the first level IIRC - it uses subdirs then.
But the root directory should not be in `$PATH` then anymore, is it?
(better to follow up / ask on the PR though)

**Comment by blueyed** [comment]
@tmzhuang 
That does not sound like it is (directly) related to this issue and it lacks information about how the virtualenv is displayed etc.  Better to file a new issue, possibly with powerline then.  Marking as off-topic.

**Comment by tmzhuang** [comment]
> Not too sure if this is related or if this is even something to do with pyenv. I'm using powerline and there's some funky stuff going on with the the way the path is displayed depending on if I have a local pyenv set.
> 
> ![image](https://user-images.githubusercontent.com/1159725/63188842-84c55980-c030-11e9-82b7-3d185b647bf3.png)
> Note the virtualenv is incorrectly displayed until a new command is made. Also cd'ing to a different directory does not update the path correctly. This is not the case if I'm not using a .python-version to start a virtualenv.
> ![image](https://user-images.githubusercontent.com/1159725/63189003-e84f8700-c030-11e9-8943-d38cf76024fd.png)
> Although it does look like there are issues updating the virtualenv as well.

I know this was marked off-topic, but I've fixed it in case anyone runs into the same issue:
It appears that pyenv-virtualenv init and powerline-daemon both make modifications to PROMP_COMMAND. I solved this issue by changing the order in which these two are initialized in my .bashrc file (pyenv-virtualenv init after powerline-daemon).

**Comment by marcusadair** [comment]
I made the mistake of installing conda with pyenv, results below. No big deal, I just wish it shims supported a user-configurable whitelist option. As it is I can't really donate my path to pyenv+conda. 

Thanks for the otherwise awesome tool, besides the conda snafu it's been great.

```
./                                   h5mkgrp*                   pycc*
../                                  h5perf_serial*             pycodestyle*
2to3*                                h5redeploy*                pydoc*
2to3-3.7*                            h5repack*                  pydoc3*
Assistant.app*                       h5repart*                  pydoc3.7*
Designer.app*                        h5stat*                    pyflakes*
Linguist.app*                        h5unjam*                   pygmentize*
activate*                            h5watch*                   pyjson5*
anaconda*                            iconv*                     pylint*
anaconda-navigator*                  icu-config*                pylupdate5*
anaconda-project*                    icuinfo*                   pyrcc5*
asadmin*                             idle*                      pyreverse*
autopoint*                           idle3*                     pytest*
binstar*                             idle3.7*                   python*
bokeh*                               imageio_download_bin*      python-config*
bsdcat*                              imageio_remove_bin*        python.app*
bsdcpio*                             infotocap*                 python3*
bsdtar*                              instance_events*           python3-config*
bundle_image*                        iptest*                    python3.7*
bzcat*                               iptest3*                   python3.7-config*
bzcmp*                               ipython*                   python3.7-gdb.py*
bzdiff*                              ipython3*                  python3.7m*
bzegrep*                             isort*                     python3.7m-config*
bzfgrep*                             isql*                      pythonw*
bzgrep*                              isympy*                    pyuic5*
bzip2*                               iusql*                     pyvenv*
bzip2recover*                        jlpm*                      pyvenv-3.7*
bzless*                              jpegtran*                  qdbusviewer.app*
bzmore*                              jsonschema*                qdoc*
c_rehash*                            jupyter*                   qgltf*
canbusutil*                          jupyter-bundlerextension*  qlalr*
captoinfo*                           jupyter-console*           qml.app*
cfadmin*                             jupyter-kernel*            qmlcachegen*
chardetect*                          jupyter-kernelspec*        qmleasing*
cjpeg*                               jupyter-lab*               qmlimportscanner*
compile_et*                          jupyter-labextension*      qmllint*
conda*                               jupyter-labhub*            qmlmin*
conda-build*                         jupyter-migrate*           qmlprofiler*
conda-convert*                       jupyter-nbconvert*         qmlscene*
conda-debug*                         jupyter-nbextension*       qmltestrunner*
conda-develop*                       jupyter-notebook*          qscxmlc*
conda-env*                           jupyter-qtconsole*         qt.conf*
conda-index*                         jupyter-run*               qtattributionsscanner*
conda-inspect*                       jupyter-serverextension*   qtdiag*
conda-metapackage*                   jupyter-troubleshoot*      qtpaths*
conda-render*                        jupyter-trust*             qtplugininfo*
conda-server*                        jupyter_mac.command*       qwebengine_convert_dict*
conda-skeleton*                      k5srvutil*                 raw2tiff*
conda-verify*                        kadmin*                    rdjpgcom*
cph*                                 kdestroy*                  recode-sr-latin*
cq*                                  keyring*                   repc*
curve_keygen*                        kill_instance*             reset*
cwutil*                              kinit*                     rg*
cygdb*                               klist*                     route53*
cython*                              kpasswd*                   rst2html.py*
cythonize*                           krb5-config*               rst2html4.py*
dask-remote*                         ksu*                       rst2html5.py*
dask-scheduler*                      kswitch*                   rst2latex.py*
dask-ssh*                            ktutil*                    rst2man.py*
dask-submit*                         kvno*                      rst2odt.py*
dask-worker*                         launch_instance*           rst2odt_prepstyles.py*
dbus-cleanup-sockets*                libpng16-config*           rst2pseudoxml.py*
dbus-daemon*                         list_instances*            rst2s5.py*
dbus-launch*                         lss3*                      rst2xetex.py*
dbus-monitor*                        lz4*                       rst2xml.py*
dbus-run-session*                    lz4c*                      rstpep2html.py*
dbus-send*                           lz4cat*                    runxlrd.py*
dbus-test-tool*                      lzcat*                     s3put*
dbus-update-activation-environment*  lzcmp*                     samp_hub*
dbus-uuidgen*                        lzdiff*                    sclient*
deactivate*                          lzegrep*                   sdbadmin*
derb*                                lzfgrep*                   showtable*
djpeg*                               lzgrep*                    sim_client*
dltest*                              lzless*                    sip*
dynamodb_dump*                       lzma*                      skivi*
dynamodb_load*                       lzmadec*                   slencheck*
easy_install*                        lzmainfo*                  sphinx-apidoc*
easy_install-3.7*                    lzmore*                    sphinx-autogen*
elbadmin*                            macchangeqt*               sphinx-build*
envsubst*                            macdeployqt*               sphinx-quickstart*
epylint*                             makeconv*                  spyder*
f2py*                                msgattrib*                 spyder3*
f2py3*                               msgcat*                    sqlite3_analyzer*
f2py3.7*                             msgcmp*                    symilar*
fax2ps*                              msgcomm*                   syncqt.pl*
fax2tiff*                            msgconv*                   tabs*
fetch_file*                          msgen*                     taskadmin*
fits2bitmap*                         msgexec*                   tclsh*
fitscheck*                           msgfilter*                 tclsh8.6*
fitsdiff*                            msgfmt*                    tic*
fitsheader*                          msggrep*                   tiff2bw*
fitsinfo*                            msginit*                   tiff2pdf*
fixqt4headers.pl*                    msgmerge*                  tiff2ps*
flask*                               msgunfmt*                  tiff2rgba*
futurize*                            msguniq*                   tiffcmp*
gdbus-codegen*                       mturk*                     tiffcp*
genbrk*                              navigator-updater*         tiffcrop*
gencfu*                              ncursesw6-config*          tiffdither*
gencnval*                            ngettext*                  tiffdump*
gendict*                             nosetests*                 tiffinfo*
genrb*                               numba*                     tiffmedian*
gettext*                             odbc_config*               tiffset*
gettext.sh*                          odbcinst*                  tiffsplit*
gettextize*                          pal2rgb*                   toe*
gif2h5*                              pandoc*                    tqdm*
gio*                                 pandoc-citeproc*           tset*
glacier*                             pasteurize*                unlz4*
glib-genmarshal*                     pcre-config*               unlzma*
glib-gettextize*                     pcregrep*                  unzstd*
glib-mkenums*                        pcretest*                  uuclient*
gobject-query*                       pep8*                      vba_extract.py*
gss-client*                          pip*                       volint*
gtester*                             pip3*                      wcslint*
gtester-report*                      pip3.7*                    wheel*
h52gif*                              pixeltool.app*             wish*
h5c++*                               pkgdata*                   wish8.6*
h5cc*                                pkginfo*                   wrjpgcom*
h5clear*                             png-fix-itxt*              xc-avoidance*
h5copy*                              pngfix*                    xgettext*
h5debug*                             ppm2tiff*                  xlwings*
h5diff*                              pt2to3*                    xml2-config.bak*
h5dump*                              ptdump*                    xmlwf*
h5fc*                                ptrepack*                  xzcmp*
h5format_convert*                    pttree*                    xzdec*
h5import*                            py.test*                   zstd*
h5jam*                               pyami_sendmail*            zstdcat*
h5ls*                                pybabel*                   zstdmt*
```

**Comment by ashwinvis** [comment]
Installing miniconda + pyenv shims simply overrides too many core utilities, which becomes unavailable like this:
```bash
$ pip install mercurial
...
Failed to build mercurial
Installing collected packages: mercurial
  Running setup.py install for mercurial ... error
    ERROR: Command errored out with exit status 1:
     command: /home/avmo/.pyenv/versions/3.8.1/bin/python3.8 -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-6hdy_9ml/mercurial/setup.py'"'"'; __file__='"'"'/tmp/pip-install-6hdy_9ml/mercurial/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record /tmp/pip-record-hx4fum2h/install-record.txt --single-version-externally-managed --compile                                                                 
         cwd: /tmp/pip-install-6hdy_9ml/mercurial/                                                                                            
    Complete output (11 lines):                                                                                                               
    running install                                                                                                                           
    running build                                                                                                                             
    running build_mo                                                                                                                          
    generating mercurial/locale/ru/LC_MESSAGES/hg.mo from i18n/ru.po                                                                          
    msgfmt -v -o mercurial/locale/ru/LC_MESSAGES/hg.mo i18n/ru.po -c                                                                          
    pyenv: msgfmt: command not found                                                                                                          
                                                                                                                                              
    The `msgfmt' command exists in these Python versions:                                                                                     
      miniconda3-latest                                                                                                                       
                                                                                                                                              
    error: command 'msgfmt' failed with exit status 127    
```
I had to install miniconda the usual way to avoid this.

**Comment by salotz** [comment]
The whole thread is a bit complex, but for a lot of use cases in this thread it seems  weird to me to get errors like:

```
$ bzip2
pyenv: bzip2: command not found

The `bzip2' command exists in these Python versions:
  miniconda3-latest/envs/env1
  miniconda3-latest/envs/env2
```

When you are in an environment that doesn't have the command? IMO `pyenv` is working way too hard here. Just silently pass through the fact the current shims don't have the executable and let the later parts of your PATH pick it up like normal.

I don't care that other envs may have this executable. Seems like an unnecessary newby convenience reminder (that has always confused me anyways) causing the problem. Maybe I don't understand the problem well enough though.

**Comment by nchammas** [comment]
> Just silently pass through the fact the current shims don't have the executable and let the later parts of your PATH pick it up like normal.

I had this same thought after encountering this issue with the AWS CLI. pyenv masks the brew-installed version of `aws` because it's also installed into one of my pyenv-managed virtual environments, even though it's not active.

The solution in this case, which was mentioned earlier in this thread, is to install [pyenv-which-ext](https://github.com/pyenv/pyenv-which-ext). This plugin modifies pyenv's behavior so that it calls the system-provided executable if a shim is registered for the program you called but the relevant pyenv environment is not active.

It doesn't address the other issues raised in this thread, like the `if hash progname` test, and it breaks an [intentional design choice](https://github.com/pyenv/pyenv/issues/1112#issuecomment-395069476) that does help in some cases. But it does help with this issue.

**Comment by salotz** [comment]
Issues are turned off for that project (pyenv-which-ext) or I would say it there but it seems like this could be a good solution if it allowed you to configure it as an environment variable e.g. `PYENV_SHADOW` that is default turned on.

After reading this thread I'm totally convinced shells are the broken pieces not pyenv...

**Comment by brandonrobertz** [comment]
The shim code needs to make sure the thing being shimmed doesn't exist on the $PATH before clobbering and exploding IMO. This is a very frustrating and confusing default feature as-is.

**Comment by deanishe** [comment]
> The shim code needs to make sure the thing being shimmed doesn't exist on the $PATH before clobbering and exploding IMO.

I agree that it's bad default behaviour, but that can be achieved via a plugin.

This issue is about the existence of shims on `$PATH` for programs that aren't runnable under the current pyenv configuration. It causes common tests like `hash progname` to erroneously pass. It makes programs look like they're available when they actually aren't, and then explode when called.

**Comment by brandonrobertz** [comment]
> > The shim code needs to make sure the thing being shimmed doesn't exist on the $PATH before clobbering and exploding IMO.
> 
> I agree that it's bad default behaviour, but that can be achieved via a plugin.
> 
> This issue is about the existence of shims on `$PATH` for programs that aren't runnable under the current pyenv configuration. It causes common tests like `hash progname` to erroneously pass. It makes programs look like they're available when they actually aren't, and then explode when called.

Yeah you're right. They shouldn't exist at all if they won't run. My proposed "solution" still breaks lots of things (like shell scripts checking the existence of program in a guard).

**Comment by native-api** [comment]
@blueyed How are we going to handle version setting with `PYENV_VERSION`? We don't have a chance to run any logic this time.

**Comment by native-api** [comment]
I've got yet another solution idea: **reduce shims to only the active ones** -- enter this mode manually and/or via flag.

This will break automatic version selection by CWD (will need to run `pyenv rehash` explicitly) but... **_do non-Pyenv-aware scripts really go around switching the versions in the course of their work?_**

**Comment by native-api** [comment]
Also found https://github.com/cxreg/smartcd which runs stuff depending on CWD.

[It uses a **shell prompt hook**](https://github.com/cxreg/smartcd/blob/82d693a0f098caabb5130edd402fda8a165337cd/lib/core/smartcd#L854-L871) rather than `cd` hook.

Again, since scripts don't usually need the versions switched automatically just from CWD changing (and if they do, they can issue a `pyenv rehash`), only doing the switching when an interactive prompt is made makes sense.

**Comment by native-api** [comment]
**With "separate shim paths per shell/version", you actually don't need shims at all!**
Because you already have those "shims" -- in `$PYENV_ROOT/versions/<version>/bin`! So you just need to add that to `PATH`, and replace it when you switch the version. If there are multiple selected versions, multiple entries need to be inserted in the order of preference.

**Actually, this is what RVM does.**
But it has trouble doing that if it's not the first on `PATH` -- because it has no indicator on PATH to show where the entries should be inserted. Pyenv does have such an indicator -- the `shims` directory.

**Comment by native-api** [comment]
**I'd like to close this issue because its primary concern has been long resolved** and move to a new one if needed (see below).

---

The primary concern is that Pyenv breaks system utilities when an alternative is installed (e.g. with Anaconda) but is inactive. This has been fixed in #1797 .

The secondary concern is that shims break program availability tests like `if which <command> &>/dev/null` that only test for executable's presence.

The only way to address the secondary concern is to either run "reduced shims" (only for the current version), or no shims at all (add `$PYENV_ROOT/versions/<version>/bin` to `PATH` instead). Either of these severely limits Pyenv's functionality (no more seamless automagic vesion switches -- the switch will need to be done manually, either by hand or with a shell hook -- i.e. only at certain moments and likely to break other stuff that uses shell hooks, e.g. RVM) -- so the best approach I see is to introduce an optional "shimless mode" of operation that's off by default.

**Since this is a significant additional complexity and functionality reduction, I'd like to reassess how pressing the issue is in this light before going that route.**

**Proponents, please state if this is still a concern enough given the fix and if you're willing to take the functionality hit.**
